### PR TITLE
STOR-1804: use new snapshot options from ClusterCSIDriver

### PR DESCRIPTION
**Pull request by RomanBednar**
## Related PRs:
- API: https://github.com/openshift/api/pull/1783
- CSO: https://github.com/openshift/cluster-storage-operator/pull/464

## Basic smoke test:

Check current vSphere config map:
```
$ oc -n openshift-cluster-csi-drivers get cm/vsphere-csi-config -o yaml
apiVersion: v1
data:
  cloud.conf: |+
    # Labels with topology values are added dynamically via operator
    [Global]
    cluster-id = rbednar-vsphere-01-cwv8p

    [VirtualCenter "vcenter.devqe.ibmc.devcluster.openshift.com"]
    insecure-flag           = true
    datacenters             = DEVQEdatacenter
    migration-datastore-url = ds:///vmfs/volumes/vsan:527320283a8c3163-2faa6dc5949a3a28/

kind: ConfigMap
metadata:
  creationTimestamp: "2024-03-06T09:46:40Z"
  name: vsphere-csi-config
  namespace: openshift-cluster-csi-drivers
  resourceVersion: "126687"
  uid: f6968303-81d8-4048-99c1-d8211363d0fa
```

Add global snapshot limit: 
```
$ oc patch clustercsidriver/csi.vsphere.vmware.com --type=merge -p '{"spec":{"driverConfig":{"vSphere":{"globalMaxSnapshotsPerBlockVolume": 10}}}}'
clustercsidriver.operator.openshift.io/csi.vsphere.vmware.com patched
```

Verify operator reflected changes in config map:
```
$ oc -n openshift-cluster-csi-drivers get cm/vsphere-csi-config -o yaml
apiVersion: v1
data:
  cloud.conf: |+
    # Labels with topology values are added dynamically via operator
    [Global]
    cluster-id = rbednar-vsphere-01-cwv8p

    [VirtualCenter "vcenter.devqe.ibmc.devcluster.openshift.com"]
    insecure-flag           = true
    datacenters             = DEVQEdatacenter
    migration-datastore-url = ds:///vmfs/volumes/vsan:527320283a8c3163-2faa6dc5949a3a28/

    [Snapshot]
    global-max-snapshots-per-block-volume = 10

kind: ConfigMap
metadata:
  creationTimestamp: "2024-03-06T09:46:40Z"
  name: vsphere-csi-config
  namespace: openshift-cluster-csi-drivers
  resourceVersion: "127118"
  uid: f6968303-81d8-4048-99c1-d8211363d0fa
```

Verify cloud.conf mounted to pod is updated correctly:
```
$ oc -n openshift-cluster-csi-drivers exec pod/vmware-vsphere-csi-driver-controller-6f88544d99-flnsw cat /etc/kubernetes/vsphere-csi-config/cloud.conf
kubectl exec [POD] [COMMAND] is DEPRECATED and will be removed in a future version. Use kubectl exec [POD] -- [COMMAND] instead.
Defaulted container "csi-driver" out of: csi-driver, driver-kube-rbac-proxy, csi-provisioner, provisioner-kube-rbac-proxy, csi-attacher, attacher-kube-rbac-proxy, csi-resizer, resizer-kube-rbac-proxy, csi-snapshotter, snapshotter-kube-rbac-proxy, csi-liveness-probe, vsphere-syncer, syncer-kube-rbac-proxy
# Labels with topology values are added dynamically via operator
[Global]
cluster-id = rbednar-vsphere-01-cwv8p

[VirtualCenter "vcenter.devqe.ibmc.devcluster.openshift.com"]
insecure-flag           = true
datacenters             = DEVQEdatacenter
migration-datastore-url = ds:///vmfs/volumes/vsan:527320283a8c3163-2faa6dc5949a3a28/

[Snapshot]
global-max-snapshots-per-block-volume = 10
```

Remove the newly added option:
```
$ oc patch clustercsidriver/csi.vsphere.vmware.com --type=json -p='[{"op": "remove", "path": "/spec/driverConfig/vSphere"}]'
clustercsidriver.operator.openshift.io/csi.vsphere.vmware.com patched
```

Verify the config map is restored:
```
$ oc -n openshift-cluster-csi-drivers get cm/vsphere-csi-config -o yaml
apiVersion: v1
data:
  cloud.conf: |+
    # Labels with topology values are added dynamically via operator
    [Global]
    cluster-id = rbednar-vsphere-01-cwv8p

    [VirtualCenter "vcenter.devqe.ibmc.devcluster.openshift.com"]
    insecure-flag           = true
    datacenters             = DEVQEdatacenter
    migration-datastore-url = ds:///vmfs/volumes/vsan:527320283a8c3163-2faa6dc5949a3a28/

kind: ConfigMap
metadata:
  creationTimestamp: "2024-03-06T09:46:40Z"
  name: vsphere-csi-config
  namespace: openshift-cluster-csi-drivers
  resourceVersion: "127692"
  uid: f6968303-81d8-4048-99c1-d8211363d0fa
```

Try to set all available config options:
```
$ oc patch clustercsidriver/csi.vsphere.vmware.com --type=merge -p '{"spec":{"driverConfig":{"vSphere":{"globalMaxSnapshotsPerBlockVolume": 10, "granularMaxSnapshotsPerBlockVolumeInVSAN": 5, "granularMaxSnapshotsPerBlockVolumeInVVOL": 3}}}}'
clustercsidriver.operator.openshift.io/csi.vsphere.vmware.com patched
```

Verify all options are set:
```
$ oc -n openshift-cluster-csi-drivers get cm/vsphere-csi-config -o yaml
apiVersion: v1
data:
  cloud.conf: |+
    # Labels with topology values are added dynamically via operator
    [Global]
    cluster-id = rbednar-vsphere-01-cwv8p

    [VirtualCenter "vcenter.devqe.ibmc.devcluster.openshift.com"]
    insecure-flag           = true
    datacenters             = DEVQEdatacenter
    migration-datastore-url = ds:///vmfs/volumes/vsan:527320283a8c3163-2faa6dc5949a3a28/

    [Snapshot]
    global-max-snapshots-per-block-volume        = 10
    granular-max-snapshots-per-block-volume-vsan = 5
    granular-max-snapshots-per-block-volume-vvol = 3

kind: ConfigMap
metadata:
  creationTimestamp: "2024-03-06T09:46:40Z"
  name: vsphere-csi-config
  namespace: openshift-cluster-csi-drivers
  resourceVersion: "137738"
```

Clean up all config options:
```
$ oc patch clustercsidriver/csi.vsphere.vmware.com --type=json -p='[{"op": "remove", "path": "/spec/driverConfig/vSphere"}]'
clustercsidriver.operator.openshift.io/csi.vsphere.vmware.com patched
```

Set global snapshot limit to 5:
```
$ oc patch clustercsidriver/csi.vsphere.vmware.com --type=merge -p '{"spec":{"driverConfig":{"vSphere":{"globalMaxSnapshotsPerBlockVolume": 5}}}}'
clustercsidriver.operator.openshift.io/csi.vsphere.vmware.com patched

```

Create 6 snapshots and verify the 6th failed to create:
```
$ oc get volumesnapshot
NAME             READYTOUSE   SOURCEPVC   SOURCESNAPSHOTCONTENT   RESTORESIZE   SNAPSHOTCLASS     SNAPSHOTCONTENT                                    CREATIONTIME   AGE
pvc-1-snapshot   true         pvc-1                               1Gi           csi-vsphere-vsc   snapcontent-24979f14-ff20-43c6-8869-c06ac1cee714   102s           107s
pvc-2-snapshot   true         pvc-1                               1Gi           csi-vsphere-vsc   snapcontent-c14c034b-e281-4292-9893-18010a28d95f   82s            90s
pvc-3-snapshot   true         pvc-1                               1Gi           csi-vsphere-vsc   snapcontent-4884d8aa-4656-4e5c-8b98-75a87f6b8c3a   73s            80s
pvc-4-snapshot   true         pvc-1                               1Gi           csi-vsphere-vsc   snapcontent-800e4957-2487-4645-8b29-97f61fe84c84   62s            68s
pvc-5-snapshot   true         pvc-1                               1Gi           csi-vsphere-vsc   snapcontent-9e753c61-bac3-4d7d-9ae0-e855089bd445   52s            60s
pvc-6-snapshot   false        pvc-1                                             csi-vsphere-vsc   snapcontent-0b1f840f-1a10-4dd8-93b9-e6e7864d0097                  40s
```

### CODE REVIEW

**Comment by RomanBednar** [comment]
Re-tested with latest openshift/api PR, looks ok:

```
$ oc patch clustercsidriver/csi.vsphere.vmware.com --type=merge -p '{"spec":{"driverConfig":{"vSphere":{"globalMaxSnapshotsPerBlockVolume": 10}}}}'
clustercsidriver.operator.openshift.io/csi.vsphere.vmware.com patched

$ oc -n openshift-cluster-csi-drivers get cm/vsphere-csi-config -o yaml
apiVersion: v1
data:
  cloud.conf: |
    # Labels with topology values are added dynamically via operator
    [Global]
    cluster-id = ci-ln-ddsrk9t-c1627-7ztq2

    [VirtualCenter "v8c-2-vcenter.ocp2.dev.cluster.com"]
    insecure-flag           = true
    datacenters             = IBMCloud
    migration-datastore-url = ds:///vmfs/volumes/vsan:52bbf3f0bbdd6a6a-81552b2d8f65444e/

    [Snapshot]
    global-max-snapshots-per-block-volume = 10
kind: ConfigMap
metadata:
  creationTimestamp: "2024-04-08T09:01:44Z"
  name: vsphere-csi-config
  namespace: openshift-cluster-csi-drivers
  resourceVersion: "44188"
  uid: 8b411f39-1205-439e-b324-195c8ce41071
```

**Comment by Phaow** [comment]
/label qe-approved
@RomanBednar Thanks for helping use the temp api version to pre merge testing. After the API PR merged we could bump the api version to the latest offical one and merge this PR.
Pre verify passed with `4.16.0-0.test-2024-04-08-003725-ci-ln-sv9g1z2-latest` (`build 4.16.0-0.ci,openshift/api#1783,openshift/vmware-vsphere-csi-driver-operator#226,openshift/cluster-storage-operator#464`)

```console
1. Check the default setting should be `3`.

# Create 4 snapshots and verify the 4th failed to create:
$ oc -n default get volumesnapshots/app-pvc-snapshot-4 -oyaml
...
 error:
    message: 'Failed to check and update snapshot content: failed to take snapshot of the volume cf20c294-9af7-4722-89fd-b95921b30df0: "rpc error: code = FailedPrecondition desc = the number of snapshots on the source volume cf20c294-9af7-4722-89fd-b95921b30df0 reaches the configured maximum (3)"'
...

2. Change the `global-max-snapshots-per-block-volume = 10` and `granular-max-snapshots-per-block-volume-vsan = 5 `, check the granular-max-snapshots-per-block-volume-vsan  should overwrite the max value for vSAN volumes as expected.
$ oc patch clustercsidriver/csi.vsphere.vmware.com --type=merge -p '{"spec":{"driverConfig":{"vSphere":{"globalMaxSnapshotsPerBlockVolume": 10}}}}'
clustercsidriver.operator.openshift.io/csi.vsphere.vmware.com patched 
$ oc patch clustercsidriver/csi.vsphere.vmware.com --type=merge -p '{"spec":{"driverConfig":{"vSphere":{"granularMaxSnapshotsPerBlockVolumeInVSAN": 5}}}}'
clustercsidriver.operator.openshift.io/csi.vsphere.vmware.com patched 

$ oc -n openshift-cluster-csi-drivers get cm/vsphere-csi-config -o yaml
apiVersion: v1
data:
  cloud.conf: |
    # Labels with topology values are added dynamically via operator
    [Global]
    cluster-id = pewang-0408vv-pf8v5

    [VirtualCenter "vcenter.devqe.ibmc.devcluster.openshift.com"]
    insecure-flag           = true
    datacenters             = DEVQEdatacenter
    migration-datastore-url = ds:///vmfs/volumes/vsan:527320283a8c3163-2faa6dc5949a3a28/

    [Snapshot]
    global-max-snapshots-per-block-volume        = 10
    granular-max-snapshots-per-block-volume-vsan = 5
kind: ConfigMap
metadata:
  creationTimestamp: "2024-04-08T01:27:45Z"
  name: vsphere-csi-config
  namespace: openshift-cluster-csi-drivers
  resourceVersion: "330478"
  uid: c6d0dac1-6ece-48d3-a0db-1932c18ff06e

# Create 6 snapshots and verify the 6th failed to create:
$ oc -n mystorage get volumesnapshots/vsan-pvc-snapshot-6 -oyaml
...
status:
  boundVolumeSnapshotContentName: snapcontent-ada27b34-f71a-42d6-8e0c-c4a4368f42c4
  error:
    message: 'Failed to check and update snapshot content: failed to take snapshot of the volume 314d491b-d972-4177-8024-04b41c9306d5: "rpc error: code = FailedPrecondition desc = the number of snapshots on the source volume 314d491b-d972-4177-8024-04b41c9306d5 reaches the configured maximum (5)"'
    time: '2024-04-08T03:27:57Z'
  readyToUse: false
...

3. Remove the configurations it should be recoverd to be the default value.

$ oc patch clustercsidriver/csi.vsphere.vmware.com --type=json -p='[{"op": "remove", "path": "/spec/driverConfig/vSphere"}]'
clustercsidriver.operator.openshift.io/csi.vsphere.vmware.com patched

$ oc -n openshift-cluster-csi-drivers get cm/vsphere-csi-config -o yaml
apiVersion: v1
data:
  cloud.conf: |
    # Labels with topology values are added dynamically via operator
    [Global]
    cluster-id = pewang-0408vv-pf8v5

    [VirtualCenter "vcenter.devqe.ibmc.devcluster.openshift.com"]
    insecure-flag           = true
    datacenters             = DEVQEdatacenter
    migration-datastore-url = ds:///vmfs/volumes/vsan:527320283a8c3163-2faa6dc5949a3a28/
kind: ConfigMap
metadata:
  creationTimestamp: "2024-04-08T01:27:45Z"
  name: vsphere-csi-config
  namespace: openshift-cluster-csi-drivers
  resourceVersion: "333640"
  uid: c6d0dac1-6ece-48d3-a0db-1932c18ff06e

# Create 4 snapshots and verify the 4th failed to create:
$ oc -n default get volumesnapshots/app-pvc-snapshot-4 -oyaml
...
  error:
    message: 'Failed to check and update snapshot content: failed to take snapshot of the volume 314d491b-d972-4177-8024-04b41c9306d5: "rpc error: code = FailedPrecondition desc = the number of snapshots on the source volume 314d491b-d972-4177-8024-04b41c9306d5 reaches the configured maximum (3)"'
    time: '2024-04-08T14:20:02Z'
  readyToUse: false
...

```

File: pkg/operator/utils/config.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+		case "granular-max-snapshots-per-block-volume-vsan":
+			value = vSphereConfig.GranularMaxSnapshotsPerBlockVolumeInVSAN
+		case "granular-max-snapshots-per-block-volume-vvol":
+			value = vSphereConfig.GranularMaxSnapshotsPerBlockVolumeInVVOL
+		}
+
+		if value != nil {
+			snapshotOptions[option] = strconv.FormatUint(uint64(*value), 10)
+		}
+	}
```
Comment (jsafrane) [inline pkg/operator/utils/config.go:37]: Why is it so complicated? Why do you iterate over an array of options to have a `case` for each of the option?!

```go
func addOption(options map[string]string, name string, intValue *uint32) {
    if stringValue == nil { return }
    options[name] = strconv.FormatUint(uint64(*value), 10)
}

func GetSnapshotOptions(clusterCSIDriver *opv1.ClusterCSIDriver) map[string]string {
    snapshotOptions := map[string]string{}
    if clusterCSIDriver == nil || clusterCSIDriver.Spec.DriverConfig.VSphere == nil { return }

    addOption(snapshotOptions, "global-max-snapshots-per-block-volume", vSphereConfig.GlobalMaxSnapshotsPerBlockVolume)
   ...
}

Comment (RomanBednar) [reply pkg/operator/utils/config.go]: Seemed slightly better than calling the function repeatedly, but this is more readable - changed.

File: pkg/operator/utils/config_test.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+				"granular-max-snapshots-per-block-volume-vsan": "10",
+				"granular-max-snapshots-per-block-volume-vvol": "15",
+			},
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			result := GetSnapshotOptions(test.clusterCSIDriver)
+			if !reflect.DeepEqual(result, test.expectedResult) {
+				t.Errorf("Unexpected result: %s\nExpectec: %s", result, test.expectedResult)
```
Comment (jsafrane) [inline pkg/operator/utils/config_test.go:71]: typo: Expectec

Comment (RomanBednar) [reply pkg/operator/utils/config_test.go]: Fixed.

Comment (RomanBednar) [reply pkg/operator/utils/config_test.go]: @jsafrane Need re-lgtm.

**Pull request merged**